pydoctor/test/test_astbuilder.py

```diff
@@ -1612,3 +1612,79 @@ def func():
     ''', systemcls=systemcls)
     outer = mod.contents['outer']
     assert not outer.contents
+
+@systemcls_param
+def test_constant_module(systemcls: Type[model.System]) -> None:
+    mod = fromText('''
+    LANG = 'FR'
+    ''', systemcls=systemcls)
+    assert getattr(mod.contents['LANG'], 'kind') == model.DocumentableKind.CONSTANT
+    assert getattr(mod.contents['LANG'], 'value') is not None
+
+@systemcls_param
+def test_constant_module_with_final(systemcls: Type[model.System]) -> None:
+    mod = fromText('''
+    from typing import Final
+    lang: Final = 'fr'
+    ''', systemcls=systemcls)
+    assert getattr(mod.contents['lang'], 'kind') == model.DocumentableKind.CONSTANT
+    assert getattr(mod.contents['lang'], 'value') is not None
+
+@systemcls_param
+def test_constant_class(systemcls: Type[model.System]) -> None:
+    mod = fromText('''
+    class Clazz:
+        """Class."""
+        LANG = 'FR'
+    ''', systemcls=systemcls)
+    assert getattr(mod.resolveName('Clazz.LANG'), 'kind') == model.DocumentableKind.CONSTANT
+    assert getattr(mod.resolveName('Clazz.LANG'), 'value') is not None
+
+
+@systemcls_param
+def test_constant_init(systemcls: Type[model.System]) -> None:
+    mod = fromText('''
+    class Clazz:
+        """Class."""
+        def __init__(**args):
+            self.LANG = 'FR'
+    ''', systemcls=systemcls)
+    assert getattr(mod.resolveName('Clazz.LANG'), 'kind') == model.DocumentableKind.CONSTANT
+    assert getattr(mod.resolveName('Clazz.LANG'), 'value') is not None
+
+@systemcls_param
+def test_constant_override_warns(systemcls: Type[model.System], capsys: CapSys) -> None:
+    mod = fromText('''
+    class Clazz:
+        """Class."""
+        LANG = 'EN'
+        def __init__(**args):
+            self.LANG = 'FR'
+    ''', systemcls=systemcls, modname="mod")
+    assert getattr(mod.resolveName('Clazz.LANG'), 'kind') == model.DocumentableKind.CONSTANT
+    val = getattr(mod.resolveName('Clazz.LANG'), 'value')
+    assert val is not None
+    assert ast.literal_eval(val) == 'FR'
+
+    captured = capsys.readouterr().out
+    assert "mod:4: Assignment to constant \"LANG\" overrides previous assignment.\n" in captured
+
+@systemcls_param
+def test_constant_override_do_not_warns_when_defined_in_docstring(systemcls: Type[model.System], capsys: CapSys) -> None:
+    mod = fromText('''
+    from enum import Enum
+    class PrivacyClass(Enum):
+        """
+        @cvar HIDDEN: Don't show the object at all.
+        @cvar PRIVATE: Show, but de-emphasize the object.
+        @cvar VISIBLE: Show the object as normal.
+        """
+        HIDDEN = 0
+        PRIVATE = 1
+        VISIBLE = 2
```

Comment:
Maybe add some diversity to the testing contants...and a "real" variable overwriting.

I expect that defining the constant multiple time can be found via an `if/else` expression.

I read this test multiple time trying to understand why we need 3 constants defined for this test... and I was not sure why a single constant definition is not enough.

```suggestion
        @cvar UPPER_CASE: Don't show the object at all.
        @cvar lower_case: Show, but de-emphasize the object.
        """
        UPPER_CASE = 0
        lower_case = 1
        UPPER_CASE = 2  # Overwrite the constant.
```

Reply:
Thanks @adiroiban and @mthuurne for the quick comments. 

I agree we don't need 3 constants for this test, I agree. A lowercase name, even in an `Enum, will not be recognized as Constant unless we add `typing.Final` annotation though.

Reply:
Also the override is tested in another test case.